pkg/domain/infra/abi/quadlet.go

```diff
@@ -0,0 +1,764 @@
+		if !ok {
+			logrus.Errorf("Unexpected unit returned by systemd - was not searching for %s", unitStatus.Name)
+		}
+
+		logrus.Debugf("Unit %s has status %s %s %s", unitStatus.Name, unitStatus.LoadState, unitStatus.ActiveState, unitStatus.SubState)
+		report.UnitName = unitStatus.Name
+
+		// Unit is not loaded
+		if unitStatus.LoadState != "loaded" {
+			report.Status = "Not loaded"
+		} else {
+			report.Status = fmt.Sprintf("%s/%s", unitStatus.ActiveState, unitStatus.SubState)
+		}
+		reports = append(reports, &report)
+		delete(partialReports, unitStatus.Name)
+	}
+
+	// This should not happen.
+	// Systemd will give us output for everything we sent to them, even if it's not a valid unit.
+	// We can find them with LoadState, as we do above.
+	// Handle it anyways because it's easy enough to do.
+	for _, report := range partialReports {
+		report.Status = "Not loaded"
+		reports = append(reports, &report)
+	}
+
+	finalReports := make([]*entities.ListQuadlet, 0, len(reports))
+	for _, report := range reports {
+		include := true
+		for _, filterFunc := range filterFuncs {
+			include = filterFunc(report)
+		}
+		if include {
+			finalReports = append(finalReports, report)
+		}
+	}
+
+	return finalReports, nil
+}
+
+// Retrieve path to a Quadlet file given full name including extension
+func getQuadletPathByName(name string) (string, error) {
+	// Check if we were given a valid extension
+	if !systemdquadlet.IsExtSupported(name) {
+		return "", fmt.Errorf("%q is not a supported quadlet file type", filepath.Ext(name))
+	}
+
+	quadletDirs := systemdquadlet.GetUnitDirs(rootless.IsRootless())
+	for _, dir := range quadletDirs {
+		testPath := filepath.Join(dir, name)
+		if _, err := os.Stat(testPath); err != nil {
+			if !errors.Is(err, fs.ErrNotExist) {
+				return "", fmt.Errorf("cannot stat quadlet at path %q: %w", testPath, err)
+			}
+			continue
+		}
+		return testPath, nil
+	}
+	return "", fmt.Errorf("could not locate quadlet %q in any supported quadlet directory", name)
+}
+
+func (ic *ContainerEngine) QuadletPrint(ctx context.Context, quadlet string) (string, error) {
+	quadletPath, err := getQuadletPathByName(quadlet)
+	if err != nil {
+		return "", err
+	}
+
+	contents, err := os.ReadFile(quadletPath)
+	if err != nil {
+		return "", fmt.Errorf("reading quadlet %q contents: %w", quadletPath, err)
+	}
+
+	return string(contents), nil
+}
+
+// QuadletRemove removes one or more Quadlet files from the system.
+// It handles both individual Quadlet files and Quadlet applications (collections of related files).
+// The function performs the following operations:
+//   - Validates that specified Quadlet files exist
+//   - Checks if associated systemd services are running
+//   - Optionally stops running services (if force option is set)
+//   - Removes the Quadlet files and any associated asset files
+//   - Reloads systemd daemon if necessary
+//
+// Parameters:
+//   - ctx: Context for cancellation and timeout handling
```

Comment:
it can be assumed that people know what a context is documenting this is just pointless noise.